File: services/location_ingest/src/freshness.ts
import type { Freshness } from "@rounds/contracts";

export type FreshnessPolicy = {
  liveThroughMs: number;
  agingThroughMs: number;
  staleThroughMs: number;
};

export const phaseZeroFreshnessPolicy: FreshnessPolicy = {
  liveThroughMs: 20_000,
  agingThroughMs: 45_000,
  staleThroughMs: 5 * 60_000,
};

export function classifyFreshness(
  sourceAt: Date | undefined,
  now: Date,
  policy: FreshnessPolicy = phaseZeroFreshnessPolicy,
): Freshness {
  if (!sourceAt) return "unknown";
  const age = Math.max(0, now.getTime() - sourceAt.getTime());
  if (age <= policy.liveThroughMs) return "live";
  if (age <= policy.agingThroughMs) return "aging";
  if (age <= policy.staleThroughMs) return "stale";
  return "unknown";
}

File: apps/driver_harness/lib/src/telemetry/freshness.dart
enum PositionFreshness { live, aging, stale, unknown }

class FreshnessPolicy {
  const FreshnessPolicy({
    this.liveThrough = const Duration(seconds: 20),
    this.agingThrough = const Duration(seconds: 45),
    this.staleThrough = const Duration(minutes: 5),
  });

  final Duration liveThrough;
  final Duration agingThrough;
  final Duration staleThrough;

  PositionFreshness classify(DateTime? sourceAt, DateTime now) {
    if (sourceAt == null) return PositionFreshness.unknown;
    final age = now.toUtc().difference(sourceAt.toUtc());
    if (age <= liveThrough) return PositionFreshness.live;
    if (age <= agingThrough) return PositionFreshness.aging;
    if (age <= staleThrough) return PositionFreshness.stale;
    return PositionFreshness.unknown;
  }
}
